modmail/dispatcher.py

```diff
@@ -0,0 +1,167 @@
+import asyncio
+import bisect
+import logging
+from typing import Callable, Coroutine, Dict, List, Optional, Union
+
+from modmail import ModmailLogger
+
+logger: ModmailLogger = logging.getLogger(__name__)
+
+CoroutineFunction = Callable[..., Coroutine]
+
+
+class Dispatcher:
+    """
+    Dispatches events through an async event handler system.
+
+    Supports blocking events and priority.
+    """
+
+    # These are separate because it makes using the bisect module easier. See _register_handler.
+    blocking_handlers: Dict[str, List[Optional[CoroutineFunction]]]
+    blocking_priorities: Dict[str, List[int]]
+    handlers: Dict[str, List[CoroutineFunction]]
+
+    def __init__(self, *event_names: str):
+        self.handlers = {}
+        self.blocking_handlers = {}
+        self.blocking_priorities = {}
+
+        self.register_event(*event_names)
```

Comment:
Is it meant to not be possible to be able to make a Dispatcher without any events at all?

I feel like this line should not exist at all, and a Dispatcher must be made separately from registering events.

Reply:
I just added it for convenience. I'm not sure exactly how we intend to initialize all this, but if we want to init it with some events present, adding the line makes that possible